### PR TITLE
Z_ISR_DECLARE: ensure proper alignment on 64-bit targets

### PR DESCRIPTION
**Pull request by npitre**
Prevent spurious unexpected alignment on 64-bit targets.

### CODE REVIEW

**Review by tyrantlizard77 — DISMISSED** [review]

File: include/sw_isr_table.h [hunk 1]
```diff
@@ -65,8 +65,8 @@ struct _isr_list {
  * and/or SW ISR tables.
  */
 #define Z_ISR_DECLARE(irq, flags, func, param) \
-	static struct _isr_list Z_GENERIC_SECTION(.intList) __used \
-		_MK_ISR_NAME(func, __COUNTER__) = \
+	static struct _isr_list Z_GENERIC_SECTION(.intList) \
+	__aligned(sizeof(void *)) __used _MK_ISR_NAME(func, __COUNTER__) = \
```
Comment (andyross) [inline include/sw_isr_table.h]: What was going wrong here?  By my read struct _isr_list includes a void* field, so should naturally have had pointer alignment already.

I worry that this might be working around a deeper problem with the linker scripts...

**Comment by andrewboie** [comment]
I'm now confused by this patch as well.
The stuff going into .intlist is stripped out of the binary, it is only used by gen_isr_tables.py to create the real IRQ tables. No runtime code will ever use these structures unless I'm mistaken.

**Comment by npitre** [comment]
On Wed, 24 Jul 2019, Andy Ross wrote:

> What was going wrong here?  By my read struct _isr_list includes a void* field, so should naturally have had pointer alignment already.

It's the same issue again with the assembler free to include a bigger 
alignment (32 by default on section switches).

**Comment by npitre** [comment]
On Wed, 24 Jul 2019, Andrew Boie wrote:

> The stuff going into .intlist is stripped out of the binary, it is only used by gen_isr_tables.py to create the real IRQ tables. No runtime code will ever use these structures unless I'm mistaken.

it's not the realtime usage but the storage that may be a problem. If 
there is more than one instance of Z_ARCH_IRQ_CONNECT() in the same 
source file then the 64-bit assembler is happy to pad between them to 
the next 32-byte boundary because of the section switch between them 
(regardless of the fact that they end up in the same output section).

Mind you, I didn't actually test if that made a difference in this 
particular case. But as I've been bitten several times by this issue 
before, I simply added this as a precaution the moment I saw 
Z_ISR_DECLARE() being used in riscv/arch.h.

**Comment by gnuless** [comment]
> If there is more than one instance of Z_ARCH_IRQ_CONNECT() in the same source file then the 64-bit assembler is happy to pad between them to the next 32-byte boundary because of the section switch between them (regardless of the fact that they end up in the same output section). 

The assembler really shouldn't do things like that.  I'm not saying that it doesn't, just that it shouldn't.

> Mind you, I didn't actually test if that made a difference in this particular case.

I hate to be a wet blanket, but I don't think we should patch things if they're not known to be problems.  Maybe open an issue to investigate if you think that's warranted.

**Comment by npitre** [comment]
On Wed, 24 Jul 2019, Charles E. Youse wrote:

> > If there is more than one instance of Z_ARCH_IRQ_CONNECT() in the same source file then the 64-bit assembler is happy to pad between them to the next 32-byte boundary because of the section switch between them (regardless of the fact that they end up in the same output section). 
> 
> The assembler really shouldn't do things like that.  I'm not saying that it doesn't, just that it shouldn't.

It does. Especially with 64-bit targets where the default section 
alignment appears to be 32 instead of 4.

> > Mind you, I didn't actually test if that made a difference in this particular case.
> 
> I hate to be a wet blanket, but I don't think we should patch things if they're not known to be problems.

This is a known problem. This is a problem that has been investigated
and patched already. I just isn't localised to one spot.

Please see commit 7daa5451cf0c for one such case where this is 
already documented.

What I don't know is whether this particular case does exhibit the 
problem right now. But even if it doesn't, it is just a matter of time 
before it does. Or if you want, I sure know how to create a problem with 
it.

**Comment by gnuless** [comment]
> This is a known problem. This is a problem that has been investigated and patched already. I just isn't localised to one spot. 

If that is the case, we should probably figure out how to fix the underlying problem rather than sprinkling band-aids everywhere, no?

**Comment by andrewboie** [comment]
>  But even if it doesn't, it is just a matter of time before it does. Or if you want, I sure know how to create a problem with it.

Given that these data structures are never used at runtime, I'd be interested to know how this is going to cause problems.

**Comment by npitre** [comment]
On Wed, 24 Jul 2019, Charles E. Youse wrote:

> > This is a known problem. This is a problem that has been investigated and patched already. I just isn't localised to one spot. 
> 
> If that is the case, we should probably figure out how to fix the underlying problem rather than sprinkling band-aids everywhere, no?

If you know how to fix it better you're welcome of course.

But I admit I failed to use the very helper I created myself for such
cases which comes with actual documentation. Revised patch is coming.

**Comment by gnuless** [comment]
> If you know how to fix it better you're welcome of course. 

That's not the way this works.

**Comment by npitre** [comment]
On Wed, 24 Jul 2019, Andrew Boie wrote:

> Given that these data structures are never used at runtime, I'd be interested to know how this is going to cause problems.

The python script consuming the data is as likely to run afoul of 
unexpected padding between multiple instances of such structures. Or is 
this macro always instantiated only once in a build?

**Review by ghost — CHANGES_REQUESTED** [review]
I'd like to see the difference in the output from the compiler for these, because if the problem is "the assembler inserts too much padding in section switches," but telling the *compiler* to align it differently (which outputs, you know, a file piped into the *assembler*) fixes the problem, then there's obviously a way to avoid the over-padding introduced by the assembler.  Perhaps that will provide us with some insight into a more general fix than playing whack-a-mole.

**Comment by npitre** [comment]
On Wed, 24 Jul 2019, Charles E. Youse wrote:

> > If you know how to fix it better you're welcome of course. 
> 
> That's not the way this works.

Look: this problem was discovered, investigated, fixed, discussed, 
and a concensus was reached before the accepted fix was committed quite 
a while ago. If you don't like how it is done now then the least you 
could do is to propose something better.

**Comment by gnuless** [comment]
> If you don't like how it is done now then the least you could do is to propose something better.

I did propose something better. Figure out how to keep the assembler from behaving badly, rather than sprinkling arcane magic spells around the source code.

**Comment by npitre** [comment]
On Wed, 24 Jul 2019, Charles E. Youse wrote:

> I'd like to see the difference in the output from the compiler for 
> these, because if the problem is "the assembler inserts too much 
> padding in section switches," but telling the *compiler* to align it 
> differently (which outputs, you know, a file piped into the 
> *assembler*) fixes the problem, then there's obviously a way to avoid 
> the over-padding introduced by the assembler.

Strictly speaking, the padding is inserted by the assembler but it is 
told so by the compiler. The nastiest part comes from the fact that the 
linker may override that padding with its own ALIGN() attribute. But it 
does remove extra padding only at the end of an input section and not in 
the middle of it. If the assembler is told to align a section X to 16 
bytes, store an object A of size 8, move to another section, then come 
back to section X with align 16 to store object B of size 8, then 
you'll get a padding of 8 between A and B and the linker may only remove 
the padding that follows B. If you store objects into section X from 
multiple source files then each of those object files will see their 
padding removed by the linker and you won't see the issue in that case.

Here's some code to demonstrate it:

```
struct foo { long a, b, c; };
struct foo A __attribute__((section("adhoc1"))) = { 1, 2, 3 };
struct foo B __attribute__((section("adhoc2"))) = { 4, 5, 6 };
struct foo C __attribute__((section("adhoc1"))) = { 7, 8, 9 };
```

The output you may expect:

```
        .globl  C
        .section        adhoc1,"aw"
        .align 16
        .type   C, @object
        .size   C, 24
C:
        .quad   7
        .quad   8
        .quad   9
        .globl  B
        .section        adhoc2,"aw"
        .align 16
        .type   B, @object
        .size   B, 24
B:
        .quad   4
        .quad   5
        .quad   6
        .globl  A
        .section        adhoc1
        .align 16
        .type   A, @object
        .size   A, 24
A:
        .quad   1
        .quad   2
        .quad   3
```

With the same change as provided in this PR, those `.align 16` become 
`.align 8`.

**Comment by npitre** [comment]
And when I say the linker removes padding at the end, that may as well 
be at the beginning too. But that's not the issue. It's the padding 
inserted between two ormore objects by the assembler to enforce the 
requested alignment that is the problem.

**Comment by gnuless** [comment]
> It's the padding inserted between two or more objects by the assembler to enforce the requested alignment that is the problem.

You keep dragging the poor assembler into this.  The assembler is doing what it's told.  Here's a better statement of the problem:

1. When GCC outputs an object to a user-defined section, it outputs an arbitrary alignment after the `.section` directive.  [arbitrary from our perspective -- not GCC's] If this alignment is too restrictive, then our attempts to construct C arrays in bits in pieces breaks.
2. The linker inadvertently masks this problem in most circumstances, when at most one element per section/array appears in any intermediate object file.  The problem disappears because the linker only enforces its own alignment restrictions, and does not (indeed, can not) enforce the `.align` directives output by the compiler.

It just so happens that for most (all?) targets until recently, (1) didn't introduce overly-strict alignments, and in most circumstances now when it does, (2) hides the problem.

While I agree that your approach is roughly the correct one (that is, suppressing GCC's overly-cautious `.align` directives), I have several problems with it:

1. This is a GCC-specific problem, and the fix(es) appear in multiple non-GCC-specific files.
2. The fix is not centralized. It appears in more than one place (and probably will appear in others), which means that one *instance* has now become a *pattern*, and we should find a way to treat this problem as a whole, rather than playing whack-a-mole with each instance.
3. The fixes are arcane and uncommented, e.g., forcing a struct to be aligned to its natural alignment is *prima facie* nonsense.  At a minimum, a comment should be here explaining the need for something that appears unnecessary.
4. The fixes are inconsistent.  In one, you align to the alignment of a type, in another, you're explicitly aligning to the *size* of a type (which is probably, in itself, not a very good idea).

I might suggest introducing a new macro in the toolchain headers, something like `Z_ARRAY_SECTION(section)` which will expand to the appropriate attributes to emit an object to a section with a predictable and workable alignment.  (An alignment of '1' would probably work -- all objects in the section will be the same size and alignment, and the linker can enforce the alignment.) Then, all elements constructed throughout the kernel using this section/array building technique should be changed to use the new macro uniformly.

**Comment by andyross** [comment]
Huh... well yeah, that's weird.  This genuinely looks like a gcc bug, and I think we should report it.  Structs with section attributes get prepended alignment bytes in the assembly that don't correspond to the struct's natural alignment.

I wonder if this is there to work around linker misbehavior or something.   Those bytes are just garbage.  Structs should appear in assembly the same way they will be linked.

As far as a Zephyr workaround, maybe we need to have a unified STRUCT_IN_SECTION (sect) macro or something that takes care of the alignment and attribute parts of the declaration in unison.

**Comment by npitre** [comment]
On Thu, 25 Jul 2019, Charles E. Youse wrote:

> I might suggest introducing a new macro in the toolchain headers, 
> something like `Z_ARRAY_SECTION(section)` which will expand to the 
> appropriate attributes to emit an object to a section with a 
> predictable and workable alignment.

I invite you to look at Z_STRUCT_SECTION_ITERABLE().

> (An alignment of '1' would probably work -- all objects in the section 
> will be the same size and alignment, and the linker can enforce the 
> alignment.)

That won't work if the compiler expects a particular alignment for some 
object access (think LDRD on ARM for example). The compiler knows what 
alignment it needs not the linker, and it may be different from one case 
to another.

> Then, all elements constructed throughout the kernel using this 
> section/array building technique should be changed to use the new 
> macro uniformly.

Indeed. And a bunch of them have been converted already.

But that would imply outlawing direct usage of 
__attribute__((section())) and proxies such as Z_GENERIC_SECTION() 
throughout. I don't know if such a change would be adviseable so close 
to the 2.0 release.

In this particular case, the section name would have to change as 
Z_STRUCT_SECTION_ITERABLE() derrives it from the structure type. Yet 
this .intList section is also used with a completely different structure 
by the x86 code and I'm far from versed enough in x86 specific low-level 
stuff to risk regressing it at this point.

Hence this very simple almost one-liner fix using Z_DECL_ALIGN() which, 
by the way, _is_ documented (albeit possibly not as clearly as you did).

**Comment by gnuless** [comment]
> That won't work if the compiler expects a particular alignment for some object access (think LDRD on ARM for example). The compiler knows what alignment it needs not the linker, and it may be different from one case to another.

That's why we tell the linker what the alignment for sections should be.  You're talking out of both sides of your mouth to some degree ... the compiler knows what the right alignment should be vs the compiler is selecting the wrong alignment.

> But that would imply outlawing direct usage of __attribute__((section())) and proxies such as Z_GENERIC_SECTION() throughout.

Direct usage *should* be outlawed.  The proxies are there to abstract the toolchain specifics.

**Comment by npitre** [comment]
On Thu, 25 Jul 2019, Charles E. Youse wrote:

> > That won't work if the compiler expects a particular alignment for some object access (think LDRD on ARM for example). The compiler knows what alignment it needs not the linker, and it may be different from one case to another.
> 
> That's why we tell the linker what the alignment for sections should 
> be.  You're talking out of both sides of your mouth to some degree ... 
> the compiler knows what the right alignment should be vs the compiler 
> is selecting the wrong alignment.

As ridicule as this may sound to you, that's a pretty good summary of 
the situation.

The compiler is using some alignment by default that sometimes breaks 
our ability to view the linked binary as an array.

But the actual alignment that should be used is known by the compiler 
not the linker.

Hence we need to ask the compiler what the proper alignment is using 
__alignof(type) and use that to tell the compiler what alignment it 
should use for the section using __attribute__((aligned(x))).

Is that a bug on the compiler's part? Maybe, maybe not. If I use 2 longs 
instead of 3, then the compiler default alignment on the section becomes 
8 as we would expect. If I use -m32 then the section alignment is always 
4 on x86 regardless of the types used in the struct (even with long 
long). On ARM (32 bits) the alignment is always 4 bytes, except when 
there is a long long in the struct at which point it becomes 8 bytes. On 
riscv32 the alignment is always 4 bytes like x86-32. On riscv64 it is 
always 8 bytes even if the struct alignment is 4.

In other words, we cannot rely on the default alignment for our purpose 
as it apparently must serve another purpose.

> Direct usage *should* be outlawed.  The proxies are there to abstract the toolchain specifics.

Agreed. But is it appropriate to do that across the board right before a 
feature freeze? Here I'm leaning towards the smallest change using the 
Z_DECL_ALIGN() proxy to minimize intrusiveness.

**Comment by gnuless** [comment]
> Is that a bug on the compiler's part?

No.  GCC is well within its rights to align the output as it sees fit.  The fact that we're playing these games with the output is on us.

> Agreed. But is it appropriate to do that across the board right before a feature freeze? Here I'm leaning towards the smallest change using the Z_DECL_ALIGN() proxy to minimize intrusiveness.

Here's the problem: this fix, both in its substance and its form, is a hack.  You spelled out the "right" way to fix this above (basically: refactor `Z_STRUCT_SECTION_ITERABLE` or some variant thereof).  The real danger here is that this patch will go in and simply become the permanent fix, just adding more technical debt to the ever-growing pile.

**Review by ghost — APPROVED** [review]
My objections are on record.  -1 lifted.

**Review by tyrantlizard77 — APPROVED** [review]
